### ISSUE TITLE
[CUDA] Add channels_last_3d support for commonly used modules

### ISSUE THREAD

**Issue by xwang233** [pytorch/pytorch#59168]
## [CUDA] Add channels_last_3d support for commonly used modules

The goal is to add `channels_last_3d`, aka NDHWC, support on CUDA devices, and improve performance on 3D model training and inference, e.g. U-Net 3D.

This task is similar to the `channels_last` (2d) support in https://github.com/pytorch/pytorch/issues/28619, but would be much less complicated. This is because most of the C++ infrastructures already have some kind of support with `at::MemoryFormat::ChannelsLast3d`. All we need to do at this moment, is

- Enable `channels_last_3d` dispatch to other library functions, e.g. cuBLAS, cuDNN.
- Write `channels_last_3d` native CUDA kernels.
- Check dispatch logic, e.g. change `input.contiguous()` to `input.contiguous(input.suggest_memory_format())`. Also need to make sure this doesn't break existing `channels_last` (2d) logic.

I've done some preliminary tests to check the existing `channels_last_3d` support in pytorch, 

<details><summary>script </summary>

<p>

```python
import torch

nets = [
    torch.nn.Conv3d(8, 4, 3),
    torch.nn.ConvTranspose3d(8, 4, 3),
    torch.nn.MaxPool3d(3),
    torch.nn.AvgPool3d(3),
    torch.nn.AdaptiveMaxPool3d(8),
    torch.nn.AdaptiveAvgPool3d(8),
    torch.nn.AdaptiveMaxPool3d(1),
    torch.nn.AdaptiveAvgPool3d(1),
    torch.nn.BatchNorm3d(8),
    torch.nn.InstanceNorm3d(8),
]

mf = torch.channels_last_3d

for net in nets:
    print('\n', net)

    a = torch.randn(4, 8, 16, 16, 16, dtype=torch.half, device='cuda').to(memory_format=mf)
    a.requires_grad_()
    net = net.to(device='cuda', memory_format=mf)

    with torch.cuda.amp.autocast():
        out = net(a)
        if not out.is_contiguous(memory_format=mf):
            print('**** output is in the wrong format ****')

        loss = out.sum()
    
    loss.backward()

    if not a.grad.is_contiguous(memory_format=mf):
        print('**** data grad is in the wrong format ****')
```

</p>
</details>

and here are our current development plans for channels_last_3d:

- [X] Enable cuDNN Convolution in `Conv3d`, `ConvTranspose3d` (#48430)
- [ ] Write native kernels for `AvgPool3d`
- [ ] Write native kernels for `MaxPool3d`
- [ ] Write native kernels for `AdaptiveAvgPool3d`
- [ ] Write native kernels for `AdaptiveMaxPool3d`
- [ ] Enable fast path for `AdaptiveAvgPool3d` when output size is 1 (similar to #43986, #53607)
- [ ] Enable fast path for `AdaptiveMaxPool3d` when output size is 1
- [x] Enable cuDNN BatchNorm3d with `CUDNN_BATCHNORM_SPATIAL_PERSISTENT` (ongoing #59129)
- [ ] Enable channels_last_3d dispatch for `InstanceNorm3d` (relevant [code](https://github.com/pytorch/pytorch/blob/ac67cda272b0856c1be3ebfebc324e0b3bab286f/aten/src/ATen/native/Normalization.cpp#L555-L557))

Further work:
- [ ] Write native kernels for `padding`
- [ ] Write native kernels for `interpolate`

Basically, a PR that enables the channels_last_3d feature needs to have 
- Unit test
- Performance and numerical accuracy check for a few shapes
- Numerical accuracy check for a large number of shapes, if needed
- Convergence check for DL model training, if needed (I personally think this only applies to the persistent cuDNN BatchNorm3d)

Contributions are welcome! Please leave a comment if you would like to see other channels_last_3d features that I've missed from the above list.

cc @VitalyFedyunin @jamesr66a @ppwwyyxx @ptrblck @ngimel

**Comment by ngimel** [comment]
Don't forget Upsample*3d (unless they are done already)

**Comment by hyperfraise** [comment]
I did some work, I'm really not sure it's what you intended (I just read this after). Am I on the right track ?

**Comment by hyperfraise** [comment]
I don't think this should be closed

**Comment by hyperfraise** [comment]
Lol still not ^^ I have multiple PRs linked to this, as I was advised to do. Maybe I can do something to disable pytorchbot closing this ?

**Comment by edomerli** [comment]
Adding a real-world example for emphasizing the importance of the `InstanceNorm3d` item still open here.

I'm training a 3D ResidualEncoderUNet ([nnU-Net](https://github.com/mic-dkfz/nnunet), the gold-standard for medical image segmentation) which uses `InstanceNorm3d` as its norm layer. I tried switching to `channels_last_3d` but the whole thing got slower:

|                           | contiguous | channels_last_3d | Δ      |
|---------------------------|------------|------------------|--------|
| forward (avg/iter)        | 88.7 ms    | 131.0 ms         | +47.7% |
| backward (avg/iter)       | 174.3 ms   | 198.0 ms         | +13.6% |
| total CUDA time (5 iters) | 1.273 s    | 1.579 s          | +24.0% |

torch.profiler shows that new aten::contiguous (89 ms/5 iters), aten::copy_ (145 ms), and cuDNN convertTensor_kernel (163 ms) calls appear that aren't there in the contiguous run. I suspect it's due to InstanceNorm3d falling back to contiguous every block and forcing a conversion before and after it.

Below it's a minimal repro on some simpler code.

<details><summary> Code </summary>

<p>

  ```python
  import platform
  import subprocess
  import torch, torch.nn as nn, time
  from torch.profiler import profile, ProfilerActivity
  
  torch.backends.cudnn.benchmark = True
  
  
  def print_env_info():
      print("=== Environment ===")
      print(f"Python:      {platform.python_version()}")
      print(f"PyTorch:     {torch.__version__}")
      print(f"CUDA (torch): {torch.version.cuda}")
      print(f"cuDNN:       {torch.backends.cudnn.version()}")
      print(f"GPU:         {torch.cuda.get_device_name(0)}")
      try:
          driver = subprocess.check_output(
              ["nvidia-smi", "--query-gpu=driver_version", "--format=csv,noheader"]
          ).decode().strip()
          print(f"NVIDIA driver: {driver}")
      except Exception:
          print("NVIDIA driver: (nvidia-smi not available)")
      print()
  
  
  class Block(nn.Module):
      def __init__(self, c, use_norm=True):
          super().__init__()
          self.conv = nn.Conv3d(c, c, 3, padding=1, bias=True)
          self.norm = nn.InstanceNorm3d(c, affine=True) if use_norm else nn.Identity()
          self.act = nn.LeakyReLU(inplace=True)
      def forward(self, x):
          return self.act(self.norm(self.conv(x)))
  
  
  def bench(memory_format, use_norm, n_blocks=8, c=32, shape=(2, 32, 64, 64, 64), iters=20):
      net = nn.Sequential(*[Block(c, use_norm) for _ in range(n_blocks)]).cuda().to(memory_format=memory_format)
      x = torch.randn(*shape, device="cuda").to(memory_format=memory_format).requires_grad_()
      for _ in range(3):
          net(x).sum().backward()
      torch.cuda.synchronize()
      t0 = time.time()
      for _ in range(iters):
          net(x).sum().backward()
      torch.cuda.synchronize()
      return (time.time() - t0) / iters * 1000
  
  
  def profile_run(memory_format, label, n_blocks=8, c=32, shape=(2, 32, 64, 64, 64), iters=5):
      net = nn.Sequential(*[Block(c, use_norm=True) for _ in range(n_blocks)]).cuda().to(memory_format=memory_format)
      x = torch.randn(*shape, device="cuda").to(memory_format=memory_format).requires_grad_()
      for _ in range(3):
          net(x).sum().backward()
      torch.cuda.synchronize()
  
      with profile(activities=[ProfilerActivity.CPU, ProfilerActivity.CUDA]) as prof:
          for _ in range(iters):
              net(x).sum().backward()
          torch.cuda.synchronize()
  
      print(f"\n=== torch.profiler: {label} ===")
      print(prof.key_averages().table(sort_by="cuda_time_total", row_limit=12))
  
  
  
  print_env_info()
  
  print("=== With InstanceNorm3d ===")
  c_norm = bench(torch.contiguous_format, use_norm=True)
  cl_norm = bench(torch.channels_last_3d, use_norm=True)
  print(f"contiguous:        {c_norm:.2f} ms/iter")
  print(f"channels_last_3d:  {cl_norm:.2f} ms/iter  ({(cl_norm/c_norm - 1)*100:+.1f}%)")
  
  print("\n=== Without InstanceNorm3d (nn.Identity instead) ===")
  c_nonorm = bench(torch.contiguous_format, use_norm=False)
  cl_nonorm = bench(torch.channels_last_3d, use_norm=False)
  print(f"contiguous:        {c_nonorm:.2f} ms/iter")
  print(f"channels_last_3d:  {cl_nonorm:.2f} ms/iter  ({(cl_nonorm/c_nonorm - 1)*100:+.1f}%)")
  
  profile_run(torch.contiguous_format, "contiguous")
  profile_run(torch.channels_last_3d, "channels_last_3d")
  ```

</p>
</details>

Output (I included a run with torch.profile since I though it might help you):

```

Activated environment: wallg
=== Environment ===
Python:      3.11.7
PyTorch:     2.12.1+cu126
CUDA (torch): 12.6
cuDNN:       91002
GPU:         NVIDIA A100-SXM-64GB
NVIDIA driver: 535.274.02

=== With InstanceNorm3d ===
contiguous:        35.02 ms/iter
channels_last_3d:  47.04 ms/iter  (+34.3%)

=== Without InstanceNorm3d (nn.Identity instead) ===
contiguous:        24.82 ms/iter
channels_last_3d:  23.64 ms/iter  (-4.8%)

=== torch.profiler: contiguous ===
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  
                                                   Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg     Self CUDA   Self CUDA %    CUDA total  CUDA time avg    # of Calls  
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  
autograd::engine::evaluate_function: ConvolutionBack...         0.19%     352.062us         5.98%      10.905ms     272.623us       0.000us         0.00%      87.745ms       2.194ms            40  
                                   ConvolutionBackward0         0.06%     104.024us         5.79%      10.553ms     263.822us       0.000us         0.00%      87.745ms       2.194ms            40  
                             aten::convolution_backward         1.44%       2.633ms         5.73%      10.449ms     261.221us      84.938ms        46.34%      87.745ms       2.194ms            40  
sm80_xmma_wgrad_implicit_gemm_indexed_tf32f32_tf32f3...         0.00%       0.000us         0.00%       0.000us       0.000us      39.591ms        21.60%      39.591ms     989.781us            40  
                                           aten::conv3d         0.16%     297.321us         3.52%       6.419ms     160.475us       0.000us         0.00%      33.074ms     826.840us            40  
                                      aten::convolution         0.31%     570.213us         3.36%       6.122ms     153.042us       0.000us         0.00%      33.074ms     826.840us            40  
                                     aten::_convolution         0.75%       1.377ms         3.04%       5.551ms     138.787us       0.000us         0.00%      33.074ms     826.840us            40  
                                aten::cudnn_convolution         0.82%       1.488ms         1.91%       3.478ms      86.960us      28.464ms        15.53%      29.193ms     729.814us            40  
sm80_xmma_dgrad_implicit_gemm_tf32f32_tf32f32_f32_nh...         0.00%       0.000us         0.00%       0.000us       0.000us      27.236ms        14.86%      27.236ms     680.908us            40  
sm80_xmma_fprop_implicit_gemm_indexed_tf32f32_tf32f3...         0.00%       0.000us         0.00%       0.000us       0.000us      23.952ms        13.07%      23.952ms     598.807us            40  
autograd::engine::evaluate_function: CudnnBatchNormB...         0.24%     429.032us         1.39%       2.531ms      63.286us       0.000us         0.00%      22.213ms     555.318us            40  
                                CudnnBatchNormBackward0         0.30%     552.320us         1.15%       2.102ms      52.561us       0.000us         0.00%      22.213ms     555.318us            40  
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  
Self CPU time total: 182.408ms
Self CUDA time total: 183.300ms


=== torch.profiler: channels_last_3d ===
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  
                                                   Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg     Self CUDA   Self CUDA %    CUDA total  CUDA time avg    # of Calls  
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  
autograd::engine::evaluate_function: ConvolutionBack...         0.17%     396.314us         9.90%      22.895ms     572.387us       0.000us         0.00%      87.805ms       2.195ms            40  
                                   ConvolutionBackward0         0.04%      95.357us         9.73%      22.499ms     562.479us       0.000us         0.00%      87.805ms       2.195ms            40  
                             aten::convolution_backward         0.85%       1.963ms         9.69%      22.404ms     560.095us      75.684ms        32.22%      87.805ms       2.195ms            40  
                                            aten::copy_         0.80%       1.855ms         6.04%      13.951ms      35.772us      83.133ms        35.39%      83.133ms     213.161us           390  
void at::native::elementwise_kernel<128, 2, at::nati...         0.00%       0.000us         0.00%       0.000us       0.000us      71.167ms        30.30%      71.167ms     263.583us           270  
                                            aten::clone         0.16%     372.586us         2.24%       5.188ms      34.585us       0.000us         0.00%      43.345ms     288.966us           150  
                                    aten::instance_norm         0.11%     245.900us         3.86%       8.930ms     223.253us       0.000us         0.00%      41.604ms       1.040ms            40  
                                       aten::contiguous         0.05%     105.542us         1.87%       4.332ms      39.386us       0.000us         0.00%      39.176ms     356.143us           110  
sm80_xmma_wgrad_implicit_gemm_indexed_tf32f32_tf32f3...         0.00%       0.000us         0.00%       0.000us       0.000us      36.977ms        15.74%      36.977ms     924.414us            40  
                                           aten::conv3d         0.03%      67.241us         3.20%       7.394ms     184.839us       0.000us         0.00%      33.405ms     835.129us            40  
                                      aten::convolution         0.10%     237.765us         3.17%       7.326ms     183.158us       0.000us         0.00%      33.405ms     835.129us            40  
                                     aten::_convolution         0.14%     313.305us         3.07%       7.089ms     177.214us       0.000us         0.00%      33.405ms     835.129us            40  
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  
Self CPU time total: 231.172ms
Self CUDA time total: 234.901ms

```

Reproduces the regression at much smaller scale on my A100. Happy to share the full profiler trace/table if useful.

Would love to help in implementing the `channels_last_3d` support for `InstanceNorm3d` but I'm not familiar with CUDA so I'll let you guys handle it :). Just wanted to say that nnUNet is used by a tons of medical AI researcher so this would be a huge deal for our community.

If you need anything from my part let me know how I can help.

cc @xwang233 @dmenig
